java/system-test/system-test-drivers/src/main/java/sleeper/systemtest/drivers/query/WebSocketQueryDriver.java

```diff
@@ -0,0 +1,84 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sleeper.systemtest.drivers.query;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import sleeper.clients.QueryWebSocketClient;
+import sleeper.configuration.properties.table.TablePropertiesProvider;
+import sleeper.core.record.Record;
+import sleeper.core.record.serialiser.RecordJSONSerDe;
+import sleeper.core.schema.Schema;
+import sleeper.query.model.Query;
+import sleeper.systemtest.dsl.instance.SystemTestInstanceContext;
+import sleeper.systemtest.dsl.query.QueryAllTablesDriver;
+import sleeper.systemtest.dsl.query.QueryAllTablesSendAndWaitDriver;
+import sleeper.systemtest.dsl.query.QuerySendAndWaitDriver;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+public class WebSocketQueryDriver implements QuerySendAndWaitDriver {
+    private static final Logger LOGGER = LoggerFactory.getLogger(WebSocketQueryDriver.class);
+
+    private final QueryWebSocketClient queryWebSocketClient;
+    private Map<String, CompletableFuture<List<String>>> future = new HashMap<>();
+    private TablePropertiesProvider tablePropertiesProvider;
+
+    public static QueryAllTablesDriver allTablesDriver(SystemTestInstanceContext instance) {
+        return new QueryAllTablesSendAndWaitDriver(instance, new WebSocketQueryDriver(instance));
+    }
+
+    public WebSocketQueryDriver(SystemTestInstanceContext instance) {
+        this.queryWebSocketClient = new QueryWebSocketClient(instance.getInstanceProperties(), instance.getTablePropertiesProvider());
+        this.tablePropertiesProvider = instance.getTablePropertiesProvider();
+    }
+
+    @Override
+    public void send(Query query) {
+        LOGGER.info("Submitting query: {}", query.getQueryId());
+        try {
+            future.put(query.getQueryId(), queryWebSocketClient.submitQuery(query));
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void waitFor(Query query) {
+        if (future.containsKey(query.getQueryId())) {
+            LOGGER.info("Waiting for query: {}", query.getQueryId());
+            future.get(query.getQueryId()).join();
+        } else {
+            LOGGER.info("Query not found");
+        }
+    }
+
+    @Override
+    public List<Record> getResults(Query query) {
+        LOGGER.info("Loading results for query: {}", query.getQueryId());
+        Schema schema = tablePropertiesProvider.getByName(query.getTableName()).getSchema();
+        RecordJSONSerDe recordSerDe = new RecordJSONSerDe(schema);
+        return queryWebSocketClient.getResults(query).stream()
+                .map(recordSerDe::fromJson)
+                .collect(Collectors.toList());
```

Comment:
This is loading all the records twice, since they're returned in the CompletableFuture.

Reply:
True, i was thinking it could be used for when other queries are run, but now we are creating a new Client every time we run a query this isnt possible. I'll make it keep hold of the list returned from the future in waitFor(Query query)